### PR TITLE
feat(clerk-js, types): Trigger UserVerification within UserProfile

### PR DESCRIPTION
**Pull request by panteliselef**
## Description

When user re-verification is required FAPI will return a specific error. The goal of this PR is to recognise that error and display the UserVerification modal. After the verification is successful we retry the original request, but if the verification is cancelled the original request will not be retried.

### How are people not affected by this
This is a pretty core change in the UserProfile component, we need to make sure that only application that have opted-in to the experimental flows will see the new behavior of UserProfile. This is ensured by many existing tests including [this one](https://github.com/clerk/javascript/blob/e578b1599451d9f2122f12d835b510b26882e839/integration/tests/user-profile.test.ts#L162) that would hang indefinitely based on the new behavior.

### Cancelled verification

https://github.com/user-attachments/assets/39cac8e7-149a-4bec-8a7d-5d3105e01b9a

### Successful verification

https://github.com/user-attachments/assets/56f54ad4-dbc2-4308-9b84-0c601420ac9a

## Checklist

- [ ] `npm test` runs as expected.
- [ ] `npm run build` runs as expected.
- [ ] (If applicable) [JSDoc comments](https://jsdoc.app/about-getting-started.html) have been added or updated for any package exports
- [ ] (If applicable) [Documentation](https://github.com/clerk/clerk-docs) has been updated

## Type of change

- [ ] 🐛 Bug fix
- [ ] 🌟 New feature
- [ ] 🔨 Breaking change
- [ ] 📖 Refactoring / dependency upgrade / documentation
- [ ] other:

### CODE REVIEW

**Comment by nikospapcom** [comment]
!preview

File: packages/clerk-js/src/ui/foundations/zIndices.ts [hunk 1]
```diff
@@ -3,5 +3,6 @@ export const zIndices = Object.freeze({
   navbar: '100',
   fab: '9000',
   modal: '10000',
-  dropdown: '11000',
+  // either do this
+  dropdown: '9900',
```
[the remark this answers is not in the document]
Comment (panteliselef) [reply packages/clerk-js/src/ui/foundations/zIndices.ts:7]: hmmm good catch, any ideas why this menu works tho ? It is still rendered in a modal
image

**Comment by octoper** [comment]
!preview

**Comment by panteliselef** [comment]
!preview

File: packages/clerk-js/src/ui/foundations/zIndices.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (panteliselef) [reply packages/clerk-js/src/ui/foundations/zIndices.ts]: @alexcarpenter this should be resolved after the introduction of `isolate: 'isolation'`

**Comment by panteliselef** [comment]
!preview

**Comment by panteliselef** [comment]
!preview

File: packages/clerk-js/src/ui/Components.tsx [hunk 2]
```diff
@@ -84,6 +84,9 @@ export type ComponentControls = {
       | 'organizationProfile'
       | 'createOrganization'
       | 'userVerification',
+    options?: {
+      notify?: boolean;
+    },
```
Comment (panteliselef) [inline packages/clerk-js/src/ui/Components.tsx:89]: This is only internal

File: .changeset/four-oranges-clap.md [hunk 3]
```diff
@@ -0,0 +1,6 @@
+---
+"@clerk/clerk-js": minor
+"@clerk/types": minor
+---
+
+Trigger UserVerification within UserProfile
```
Comment (LekoArts) [inline .changeset/four-oranges-clap.md:6]: Let's rewrite the changelog to specify for which sections (e.g. 2FA) this new opt-in behavior can be used

File: packages/clerk-js/src/ui/Components.tsx [hunk 4]
```diff
@@ -245,13 +248,54 @@ const Components = (props: ComponentsProps) => {
...
 
-    componentsControls.closeModal = name => {
+    componentsControls.closeModal = (name, options = {}) => {
+      const { notify = true } = options;
       clearUrlStateParam();
-      setState(s => ({ ...s, [name + 'Modal']: null }));
+      setState(s => {
+        function handleCloseModalForExperimentalUserVerification() {
+          // @ts-ignore Expected value s['userVerificationModal']
+          const modal = s[name + 'Modal'] || {};
```
Comment (LekoArts) [inline packages/clerk-js/src/ui/Components.tsx:257]: Would ``s[`${name}Modal`]`` work for TS interpolation (and ts-ignore can be removed)?

File: packages/clerk-js/src/ui/hooks/useAssurance.ts [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+  metadata?: Metadata,
+  as?: 'response',
+): AssuranceHint | Response {
+  const obj = {
+    clerk_error: {
+      type: 'forbidden',
+      reason: 'assurance',
+      metadata,
+    },
+  } as AssuranceHint;
```
Comment (LekoArts) [inline packages/clerk-js/src/ui/hooks/useAssurance.ts:31]: Why not add the `metadata` to the `AssuranceHint`, then you don't need to typecast

File: packages/clerk-js/src/ui/hooks/useAssurance.ts [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+
+    // rethrow
+    throw e;
+  });
+}
+
+/**
+ * Polyfill for Promise.withResolvers()
+ */
+function customPromiseWithResolves() {
```
Comment (LekoArts) [inline packages/clerk-js/src/ui/hooks/useAssurance.ts:55]: Typo

```suggestion
function customPromiseWithResolvers() {
```

File: packages/clerk-js/src/ui/Components.tsx (hunk 4 above)
Comment (panteliselef) [reply packages/clerk-js/src/ui/Components.tsx:257]: Worked, thanks !

File: packages/clerk-js/src/ui/hooks/useAssurance.ts (hunk 5 above)
Comment (panteliselef) [reply packages/clerk-js/src/ui/hooks/useAssurance.ts:31]: I may end up changing this in a followup iteration.

**Comment by panteliselef** [comment]
!snapshot

**Comment by clerk-cookie** [comment]
Hey @panteliselef - the snapshot version command generated the following package versions:
| Package | Version |
| --- | --- |
| @clerk/astro | 1.3.9-snapshot.vbe1abf5 |
| @clerk/backend | 1.13.5-snapshot.vbe1abf5 |
| @clerk/chrome-extension | 1.3.11-snapshot.vbe1abf5 |
| @clerk/clerk-js | 5.24.0-snapshot.vbe1abf5 |
| @clerk/elements | 0.15.7-snapshot.vbe1abf5 |
| @clerk/clerk-expo | 2.2.17-snapshot.vbe1abf5 |
| @clerk/express | 1.0.2-snapshot.vbe1abf5 |
| @clerk/fastify | 1.0.48-snapshot.vbe1abf5 |
| @clerk/localizations | 3.1.0-snapshot.vbe1abf5 |
| @clerk/nextjs | 5.6.3-snapshot.vbe1abf5 |
| @clerk/clerk-react | 5.9.4-snapshot.vbe1abf5 |
| @clerk/remix | 4.2.32-snapshot.vbe1abf5 |
| @clerk/clerk-sdk-node | 5.0.45-snapshot.vbe1abf5 |
| @clerk/shared | 2.8.4-snapshot.vbe1abf5 |
| @clerk/tanstack-start | 0.4.8-snapshot.vbe1abf5 |
| @clerk/testing | 1.3.6-snapshot.vbe1abf5 |
| @clerk/themes | 2.1.33-snapshot.vbe1abf5 |
| @clerk/types | 4.23.0-snapshot.vbe1abf5 |

Tip: Use the snippet copy button below to quickly install the required packages.
`@clerk/astro`
```sh
npm i @clerk/astro@1.3.9-snapshot.vbe1abf5 --save-exact
```
`@clerk/backend`
```sh
npm i @clerk/backend@1.13.5-snapshot.vbe1abf5 --save-exact
```
`@clerk/chrome-extension`
```sh
npm i @clerk/chrome-extension@1.3.11-snapshot.vbe1abf5 --save-exact
```
`@clerk/clerk-js`
```sh
npm i @clerk/clerk-js@5.24.0-snapshot.vbe1abf5 --save-exact
```
`@clerk/elements`
```sh
npm i @clerk/elements@0.15.7-snapshot.vbe1abf5 --save-exact
```
`@clerk/clerk-expo`
```sh
npm i @clerk/clerk-expo@2.2.17-snapshot.vbe1abf5 --save-exact
```
`@clerk/express`
```sh
npm i @clerk/express@1.0.2-snapshot.vbe1abf5 --save-exact
```
`@clerk/fastify`
```sh
npm i @clerk/fastify@1.0.48-snapshot.vbe1abf5 --save-exact
```
`@clerk/localizations`
```sh
npm i @clerk/localizations@3.1.0-snapshot.vbe1abf5 --save-exact
```
`@clerk/nextjs`
```sh
npm i @clerk/nextjs@5.6.3-snapshot.vbe1abf5 --save-exact
```
`@clerk/clerk-react`
```sh
npm i @clerk/clerk-react@5.9.4-snapshot.vbe1abf5 --save-exact
```
`@clerk/remix`
```sh
npm i @clerk/remix@4.2.32-snapshot.vbe1abf5 --save-exact
```
`@clerk/clerk-sdk-node`
```sh
npm i @clerk/clerk-sdk-node@5.0.45-snapshot.vbe1abf5 --save-exact
```
`@clerk/shared`
```sh
npm i @clerk/shared@2.8.4-snapshot.vbe1abf5 --save-exact
```
`@clerk/tanstack-start`
```sh
npm i @clerk/tanstack-start@0.4.8-snapshot.vbe1abf5 --save-exact
```
`@clerk/testing`
```sh
npm i @clerk/testing@1.3.6-snapshot.vbe1abf5 --save-exact
```
`@clerk/themes`
```sh
npm i @clerk/themes@2.1.33-snapshot.vbe1abf5 --save-exact
```
`@clerk/types`
```sh
npm i @clerk/types@4.23.0-snapshot.vbe1abf5 --save-exact
```

**Review by alexcarpenter — CHANGES_REQUESTED** [review]
Hitting the back button has no effect

https://github.com/user-attachments/assets/ee61c6fe-0b5c-4d71-8bb0-11e65de0852b

Would be nice if there was only one overlay instead of the stacking effect where it gets darker. non-blocker though.

File: packages/clerk-js/src/ui/components/UserProfile/ActiveDevicesSection.tsx [hunk 7]
```diff
@@ -48,12 +49,19 @@ export const ActiveDevicesSection = () => {
...
+
   const revoke = async () => {
     if (isCurrent || !session) {
       return;
     }
     status.setLoading();
-    return session.revoke().finally(() => status.setIdle());
+    return (
+      handleAssurance(() => session.revoke())
+        // TODO-STEPUP: Properly handler the response with a setCardError
```
[the remark this answers is not in the document]
Comment (octoper) [reply packages/clerk-js/src/ui/components/UserProfile/ActiveDevicesSection.tsx:61]: Shouldn't this be handled everywhere then, as if the assurance fails completely we will have to show a message everywhere.

File: .changeset/four-oranges-clap.md [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+| Remove identification | ✅              | Strongest available | 10m |
+| Change primary identification | ✅              | Strongest available | 10m |
+| Update Passkey name | ❌              |  |  |
+| Enable MFA (TOTP, Phone number) | ✅              | Strongest available | 10m |
+| Disable MFA (TOΤP, Phone number) | ✅              | Strongest available | 10m |
+| Create/Regenerate Backup Codes | ✅              | Strongest available | 10m |
+| Connect External Account | ✅              | Strongest available | 10m |
+| Re-authorize External Account | ❌              |  |  |
+| Remove External Account | ✅              | Strongest available | 10m |
+| Leave organization | ❌              |  |  |
```
Comment (anagstef) [inline .changeset/four-oranges-clap.md:31]: Is this table also going into our docs?

**Review by octoper — APPROVED** [review]
Except from the comments Stefanos shared above this looks good 🔥

**Review by anagstef — APPROVED** [review]
Great work! 💯

File: .changeset/four-oranges-clap.md (hunk 8 above)
Comment (panteliselef) [reply .changeset/four-oranges-clap.md:31]: Eventually, yes we currently have it in our DX gudie

File: packages/clerk-js/src/ui/components/UserProfile/ActiveDevicesSection.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (panteliselef) [reply packages/clerk-js/src/ui/components/UserProfile/ActiveDevicesSection.tsx:61]: Actually this is not step-up specific, just something that i discovered (for later)

File: packages/clerk-js/src/ui/components/UserProfile/AccountPage.tsx [hunk 9]
```diff
@@ -53,6 +53,8 @@ export const AccountPage = withCardStateProvider(() => {
         {showEmail && <EmailsSection shouldAllowCreation={shouldAllowIdentificationCreation} />}
         {showPhone && <PhoneSection shouldAllowCreation={shouldAllowIdentificationCreation} />}
         {showConnectedAccounts && <ConnectedAccountsSection shouldAllowCreation={shouldAllowIdentificationCreation} />}
+
+        {/*TODO-STEP-UP: Verify that these work as expected*/}
```
[the remark this answers is not in the document]
Comment (panteliselef) [reply packages/clerk-js/src/ui/components/UserProfile/AccountPage.tsx:57]: Yes, although thorough testing will still need to be done. E2E tests will also be introduced before this goes to public beta

**Pull request merged**